pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1449,6 +1450,23 @@ private ByteBuf processMessageChunk(ByteBuf compressedPayload, MessageMetadata m
         // discard message if chunk is out-of-order
         if (chunkedMsgCtx == null || chunkedMsgCtx.chunkedMsgBuffer == null
                 || msgMetadata.getChunkId() != (chunkedMsgCtx.lastChunkedMessageId + 1)) {
+            // Filter duplicated chunks instead of discard it.
+            if (chunkedMsgCtx == null || msgMetadata.getChunkId() <= chunkedMsgCtx.lastChunkedMessageId) {
+                log.warn("[{}] Receive a repeated chunk messageId {}, last-chunk-id{}, chunkId = {}",
+                        msgMetadata.getProducerName(), chunkedMsgCtx == null ? null
+                                : chunkedMsgCtx.lastChunkedMessageId, msgId, msgMetadata.getChunkId());
+                compressedPayload.release();
+                increaseAvailablePermits(cnx);
+                if (chunkedMsgCtx != null) {
```

Comment:
Sure, of course.

Reply:
>It seems to be inefficient to iterate all chunks every time. Can we optimize it? I believe all chunk message ids(ledger and entry) for the same message should be the same, aren't they? Can't we check the last chunk's messageId only?

In fact, all the chunks in a chunk message are different. I also just learned it.

>It seems like the processMessageChunk does not have the id check logic. Why are we introducing this check in this PR?

Because we should check whether it is a duplicated chunk persistent in the topic or received twice by the consumer
For example:
**Case 1,  duplicated chunk persistent in the topic:**
1: uuid=p-0, mid:1:1, chunk 1 sequence ID: 0, chunk ID: 0
2: uuid=p-0, mid:1:2, chunk 2 sequence ID: 0, chunk ID: 1
3: uuid=p-0, mid:1:3, chunk 1 sequence ID: 0, chunk ID: 0 // should be acked
3: uuid=p-0, mid:1:4, chunk 2 sequence ID: 0, chunk ID: 1 // should be acked
**Case 2, received twice by the consumer:**
1: uuid=p-0, mid:1:1, chunk 1 sequence ID: 0, chunk ID: 0
2: uuid=p-0, mid:1:2, chunk 2 sequence ID: 0, chunk ID: 1
3: uuid=p-0, mid:1:1, chunk 1 sequence ID: 0, chunk ID: 0 // Just ignore it
3: uuid=p-0, mid:1:2, chunk 2 sequence ID: 0, chunk ID: 1 // Just ignore it